macaroonbakery/bakery.py

```diff
@@ -0,0 +1,280 @@
+            webbrowser.open(visit_url, new=1)
+
+            # Wait on the wait url and then get a macaroon if validated.
+            return self._acquire_macaroon_from_wait(wait_url)
+        else:
+            raise DischargeError(
+                'Unknown status code from discharge endpoint: {}'.format(
+                    response.status_code)
+            )
+
+    def _extract_macaroon_from_response(self, response):
+        ''' Extarct the macaroon from a direct successful discharge.
+
+        @param response from direct successful discharge.
+        @return a macaroon object.
+        @raises DischargeError if any error happens.
+        '''
+        try:
+            response_json = response.json()
+        except ValueError:
+            raise DischargeError(
+                'Unable to access data from discharge endpoint on 200')
+        try:
+            return deserialize(response_json['Macaroon'])
+        except KeyError:
+            raise DischargeError(
+                'Unable to access macaroon from discharge endpoint on 200')
+
+    def _acquire_macaroon_from_wait(self, wait_url):
+        ''' Wait that the user did validate its identity as the get will block
+            until then.
+            If validated then we get the macaroon from the wait endpoint
+            response.
+
+        @param wait_url the get url to call to get a macaroon.
+        @return a macaroon object
+        @raises DischargeError if any error happens.
+        '''
+
+        try:
+            resp = requests.get(wait_url, timeout=self._timeout)
+        except requests.RequestException as exc:
+            raise DischargeError(
+                'Unable to access wait url: {}'.format(exc.message))
+        if not resp.ok:
+            raise DischargeError(
+                'Unexpected status code from wait url: {}'.format(
+                    resp.status_code
+                ))
+        try:
+            response_json = resp.json()
+        except ValueError:
+            raise DischargeError(
+                'Unable to access json macaroon from wait url'
+            )
+        try:
+            discharge_token = response_json['DischargeToken']
+            macaroon = response_json['Macaroon']
+        except KeyError:
+            raise DischargeError(
+                'Unable to access macaroon or discharge token from'
+                ' discharge endpoint'
+            )
+        if self._discharge_token is None:
+            self._discharge_token = to_base64(discharge_token)
+        return deserialize(macaroon)
+
+    def _extract_urls(self, response):
+        ''' Return the visit and wait URL from response.
+
+        @param response the response from the discharge endpoint.
+        @return the visit and wait URL.
+        @raises DischargeError for ant error during the process response.
+        '''
+        try:
+            response_json = response.json()
+        except ValueError:
```

Comment:
There seems to be quite a lot of repetition of these lines. Maybe factor out some code to extract a JSON response and some keys in it?